resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c

```diff
@@ -0,0 +1,380 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "netty_window_dns_native.h"
+
+#include <winsock2.h>
+#include <Ws2ipdef.h>
+#include <ws2tcpip.h>
+#include <iphlpapi.h>
+
+#pragma comment(lib, "IPHLPAPI.lib")
+#pragma comment(lib, "ws2_32.lib")
+
+#define NETTY_JNI_UTIL_JNI_VERSION JNI_VERSION_1_6
+
+jclass nativeExceptionClass;
+
+jclass networkAdapterClass;
+jmethodID networkAdapterCtor;
+
+jclass stringClass;
+
+jclass linkedListClass;
+jmethodID linkedListCtor; 
+jmethodID linkedListAdd;
+
+jclass inetSocketAddressClass;
+jmethodID inetSocketAddressCtor;
+
+int throwNativeException(JNIEnv* env, const char* fmt, ...) {
+    va_list countArgs;
+    va_start(countArgs, fmt);
+
+    va_list printArgs;
+    va_copy(printArgs, countArgs);
+    
+    size_t length = vsnprintf(NULL, 0, fmt, countArgs);
+    size_t nbytes = length + 1;
+
+    va_end(countArgs);
+
+    char* buffer = malloc(nbytes);
+
+    if (buffer == NULL) {
+        va_end(printArgs);
+        return JNI_ERR;
+    }
+
+    vsprintf_s(buffer, nbytes, fmt, printArgs);
+    va_end(printArgs);
+
+    jint result = (*env)->ThrowNew(env, nativeExceptionClass, buffer);
+
+    free(buffer);
+
+    return result;
+}
+
+jint read_adapter_addresses(JNIEnv* env, IP_ADAPTER_ADDRESSES** target) {
+    const int MAX_TRIES = 3;
+
+    // Allocate a 15 KB buffer to start with according to https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses
+    unsigned long outBufLen = 15000;
+
+    IP_ADAPTER_ADDRESSES* pAddresses = NULL;
+    unsigned long dwRetVal = 1;
+
+    for (int tries = 0; tries < MAX_TRIES; ++tries) {
+        pAddresses = malloc(outBufLen);
+
+        if (pAddresses == NULL) {
+            fprintf(stderr, "FATAL: Out of memory.\n");
+            fflush(stderr);
+            throwNativeException(env, "Out of memory for IP_ADAPTER_ADDRESSES buffer. Tried to allocate %lu bytes.", outBufLen);
+            return JNI_ERR;
+        }
+
+        unsigned long flags = GAA_FLAG_SKIP_UNICAST | GAA_FLAG_SKIP_ANYCAST
+            | GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_FRIENDLY_NAME;
+
+        // This call can update the outBufLen variable with the actual (required) size.
+        dwRetVal = GetAdaptersAddresses(AF_UNSPEC, flags, NULL, pAddresses, &outBufLen);
+
+        if (dwRetVal != ERROR_BUFFER_OVERFLOW) {
+            break;
+        }
+
+        free(pAddresses);
+        pAddresses = NULL;
+    }
+
+    if (dwRetVal == NO_ERROR) {
+        *target = pAddresses;
+        return JNI_OK;
+    }
+
+    if (pAddresses != NULL) {
+        free(pAddresses);
+    }
```

Comment:
That won't work. In the successful case the caller has to free the IP_ADAPTER_ADDRESSES structure. Otherwise the target Variable would point to already freed memory.